src/fsharp/vs/service.fs

```diff
@@ -2215,11 +2233,36 @@ type BackgroundCompiler(referenceResolver, projectCacheSize, keepAssemblyContent
 
     // Also keyed on source. This can only be out of date if the antecedent is out of date
     let parseAndCheckFileInProjectCache = 
-        MruCache<string * string * FSharpProjectOptions, FSharpParseFileResults * FSharpCheckFileResults * int * DateTime>
+        MruCache<FileName * Source * FSharpProjectOptions, FSharpParseFileResults * FSharpCheckFileResults * FileVersion * DateTime>
             (keepStrongly=incrementalTypeCheckCacheSize,
              areSame=AreSameForChecking3,
              areSameForSubsumption=AreSubsumable3)
 
+    /// Holds keys for files being currently checked. It's used to prevent checking same file in parallel (interliveing chunck queued to Reactor).
+    let beingCheckedFileCache = 
+        System.Collections.Concurrent.ConcurrentDictionary<FilePath * ProjectPath * FileVersion, unit>(
+            { new IEqualityComparer<FilePath * ProjectPath * FileVersion> with
+                member __.Equals((f1, p1, v1), (f2, p2, v2)) = f1 = f2 && p1 = p2 && v1 = v2
+                member __.GetHashCode((f, p, v)) =
+                    let mutable hash = 17
+                    hash <- hash * 23 + f.GetHashCode()
+                    hash <- hash * 23 + p.GetHashCode()
+                    hash <- hash * 23 + v.GetHashCode()
+                    hash
+            })
+
+    /// Ensure that `f` is not called in parallel for given (file * project * version).
+    let withFileCheckLock(key: FilePath * ProjectPath * FileVersion) (f: Async<'T>) : Async<'T> =
+        async {
+            while not (beingCheckedFileCache.TryAdd(key, ())) do
+                do! Async.Sleep 100
```

Comment:
Please add some sort of limitation to this loop to prevent infinite loop should something go wrong and the key slot never become available.  I'm not sure what exactly - perhaps just a simple counter that proceeds after 100 iterations, but we should add _something_...